crates/wysiwyg/src/dom/parser/parse.rs

```diff
@@ -12,256 +12,513 @@
-            }
-            "a" => {
-                node.append_child(new_link(child));
-                convert_children(padom, child, node.last_child_mut());
-            }
-            "html" => {
-                // Skip the html tag - add its children to the
-                // current node directly.
-                convert(padom, child, node);
-            }
-            _ => {
-                // Ignore tags we don't recognise
-                // We should log - see internal task PSU-741
-            }
-        };
+        let document = parser
+            .parse_from_string(html, SupportedType::TextHtml)
+            .map_err(|_| {
+                to_dom_creation_error(
+                    "Failed to convert the Web `Document` to internal `Dom`",
+                )
+            })?;
+
+        webdom_to_dom(document).map_err(to_dom_creation_error)
     }
 
-    /// Copy all panode's information into node.
-    fn convert<S>(
-        padom: &PaDom,
-        panode: &PaNodeContainer,
-        node: &mut ContainerNode<S>,
-    ) where
+    fn webdom_to_dom<S>(webdoc: Document) -> Result<Dom<S>, Error>
+    where
         S: UnicodeString,
     {
-        for child_handle in &panode.children {
-            let child = padom.get_node(child_handle);
-            match child {
-                PaDomNode::Container(child) => {
-                    convert_container(padom, child, node);
-                }
-                PaDomNode::Document(_) => {
-                    panic!("Found a document inside a document!")
-                }
-                PaDomNode::Text(text) => {
-                    node.append_child(DomNode::new_text(
-                        text.content.as_str().into(),
-                    ));
-                }
-            }
+        let body = webdoc.body().ok_or_else(|| Error::NoBody)?;
+
+        fn convert<S>(nodes: NodeList) -> Result<Dom<S>, Error>
+        where
+            S: UnicodeString,
+        {
+            let number_of_nodes = nodes.length() as usize;
+            let mut dom = Dom::new(Vec::with_capacity(number_of_nodes));
+            let dom_document = dom.document_mut();
+
+            convert_container(nodes, dom_document)?;
+
+            Ok(dom)
         }
-    }
 
-    let mut ret = Dom::new(Vec::new());
-    let doc = ret.document_mut();
+        fn convert_container<S>(
+            nodes: NodeList,
+            dom: &mut ContainerNode<S>,
+        ) -> Result<(), Error>
+        where
+            S: UnicodeString,
+        {
+            let number_of_nodes = nodes.length() as usize;
 
-    if let PaDomNode::Document(padoc) = padom.get_document() {
-        convert(&padom, padoc, doc)
-    } else {
-        panic!("Document was not a document!");
-    }
+            for nth in 0..number_of_nodes {
+                let node = nodes.get(nth as _).unwrap();
 
-    ret
-}
+                match node.node_name().as_str() {
+                    "BR" => {
```

Comment:
OK, this is much better than I thought, because we retain fine-grained control over which tags we accept and how we convert them into Dom objects.

Reply:
Exactly. It maps to what is already done with `html5ever`.